ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Ledger/Block.hs

```diff
@@ -138,6 +140,18 @@ mkShelleyBlock raw = ShelleyBlock {
     , shelleyBlockHeaderHash = pHeaderHash $ SL.bheader raw
     }
 
+class
+  ( ShelleyBasedEra (ShelleyBlockLedgerEra blk)
+  , blk ~ ShelleyBlock (BlockProtocol blk) (ShelleyBlockLedgerEra blk)
+  ) => ShelleyBasedBlock blk
+
+instance ( proto ~ BlockProtocol (ShelleyBlock proto era)
+         , ShelleyBasedEra era
+         ) => ShelleyBasedBlock (ShelleyBlock proto era)
```

Comment:
We could give this class "more power", ie make its superclass constraints stronger, eg to `ShelleyCompatible` instead of just `ShelleyBasedEra`:
```suggestion
class
  ( ShelleyCompatible (BlockProtocol blk) (ShelleyBlockLedgerEra blk)
  , blk ~ ShelleyBlock (BlockProtocol blk) (ShelleyBlockLedgerEra blk)
  ) => ShelleyBasedBlock blk

instance ( proto ~ BlockProtocol (ShelleyBlock proto era)
         , ShelleyCompatible proto era
         ) => ShelleyBasedBlock (ShelleyBlock proto era)
```
This might come in handy in a future use case where we want to call a function in an SOP combinators that requires `ShelleyCompatible` and not just `ShelleyBasedEra`.

Reply:
I also proposed the same. I do wonder if this is not equivalent to requiring it in the constructors? as in:
```
data ShelleyBlock proto era where
  ShelleyBlock :: ShelleyCompatible proto era => 
    { ... } -> ShelleyBlock proto era
```

leaving that aside, it might be intereseting to provide instances for `Header (ShelleyBlock ...)` and `LedgerState (ShelleyBlock ...)`? I feel like this might simplify some aux classes we might have around.

Reply:
I think one should prefer function constraints, not constraints on constructors. They are not entirely equivalent

Most of the places where I use constraints inside constructors is when dealing with existentials or `Dict`s